### PR TITLE
Add buffer stream

### PR DESCRIPTION
**Pull request by A2va**
* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

File: src/tbox/stream/impl/stream/buffer.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return_val(stream_buffer && stream_buffer->buffer, -1);
+
+    // check
+    tb_check_return_val(data, -1);
+    tb_check_return_val(size, 0);
+
+    // writ data
+    if (size) tb_buffer_memncpyp(stream_buffer->buffer, stream_buffer->head, data, size);
```
Comment (A2va) [inline src/tbox/stream/impl/stream/buffer.c:128]: I noticed that the xml writer write one byte to do the formatting. Is the buffer efficient in being reallocated regularly? If not maybe we could resize the buffer to contains more than what is it written to it, a bit like a vector in C++

Comment (waruqi) [reply src/tbox/stream/impl/stream/buffer.c:128]: not need, stream has write cache. https://github.com/tboox/tbox/blob/3fdd8f0ab91cea9fe516ab31035ea79425c04c4a/src/tbox/stream/stream.c#L917

**Comment by waruqi** [comment]
error: buffer.c
src\tbox\stream\impl\stream\buffer.c(143): error C2220: the following warning is treated as an error
src\tbox\stream\impl\stream\buffer.c(143): warning C4244: '=': conversion from 'tb_hize_t' to 'tb_size_t', possible loss of data

**Comment by waruqi** [comment]
tb_buffer_memncpyp 是覆盖重写 buffer 所以每次调用，都是会重新计算当前的实际大小

这里得用 tb_buffer_resize + tb_memcpy 来做。你可以提 pr 改进下

**Pull request merged**